File: java/src/com/company/ReverseLinkedList.java
package com.company;

import java.util.List;

public class ReverseLinkedList {

    public static void main(String[] args) {
        ListNode head = ListNode.buildLinkedList(new int[]{1, 2, 3, 4, 5});
        var exp = new ReverseLinkedList();

        ListNode listNode = exp.reverseList(head);
        assert 5 == listNode.val;
        assert 4 == listNode.next.val;
        assert 3 == listNode.next.next.val;
        assert 2 == listNode.next.next.next.val;
        assert 1 == listNode.next.next.next.next.val;
        assert null == listNode.next.next.next.next.next;
        System.out.println("tests passed");
    }

    // Time O(n) and Space O(1)
    public ListNode reverseList(ListNode head) {
        if (head == null) return null;
        ListNode dummy = head;
        ListNode cur = head.next;
        if (cur == null) return dummy;
        dummy.next = null;
        while (cur != null) {
            ListNode tmp = cur;
            cur = cur.next;
            tmp.next = dummy;
            dummy = tmp;
        }

        return dummy;
    }
}

File: python/tasks/reverse_linked_list.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None

    def __str__(self):
        return f"Node({self.val})"

n1 = Node(1)
n2 = Node(2)
n3 = Node(3)
n4 = Node(4)
n5 = Node(5)

n1.next = n2
n2.next = n3
n3.next = n4
n4.next = n5

# Time O(n) and Space O(1)
def reverse_linked_list(node: Node):
        buffer = [None, node, node.next]

        while buffer[2] is not None:
            child = buffer[0]
            parent = buffer[1]
            parent.next = child

            buffer[0], buffer[1] = buffer[1], buffer[0]
            buffer[1] = buffer[2]
            buffer[2] = buffer[2].next

        buffer[1].next = buffer[0]
        return buffer[1]


r = reverse_linked_list(n1)
print(r)